### PR TITLE
PORTALS-2878

### PR DESCRIPTION
**Pull request by jay-hodgson**

### CODE REVIEW

File: packages/synapse-react-client/src/components/GenericCard/Linkify.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    return `https://pubmed.ncbi.nlm.nih.gov/${id}`
+  },
+}
+const mutationIdRule: LinkifyRule = {
+  regex: /(MGI:\d+)/,
+  onMatch: value => `https://www.informatics.jax.org/allele/${value}`,
+}
+const sciCrunchResolverRule: LinkifyRule = {
+  regex:
+    /(AB_\d+)|(Addgene_\d+)|(CVCL_[a-zA-Z0-9]+)|(MMRRC_\d+-.{3})|(IMSR_JAX:\d+)/,
```
Comment (jay-hodgson) [inline packages/synapse-react-client/src/components/GenericCard/Linkify.tsx:34]: union of all scicrunch resolver patterns

File: packages/synapse-react-client/src/components/QueryWrapperPlotNav/QueryWrapperPlotNav.stories.tsx [hunk 2]
```diff
@@ -56,11 +56,11 @@ const handlePlotClick = (event: QueryWrapperSynapsePlotRowClickEvent) => {
 export const Cards: Story = {
   args: {
     name: 'Tools',
-    sql: 'SELECT * FROM syn26438037',
+    sql: 'SELECT * FROM syn51730943',
```
Comment (jay-hodgson) [inline packages/synapse-react-client/src/components/QueryWrapperPlotNav/QueryWrapperPlotNav.stories.tsx:59]: This was failing, so updated to the new view (which also has a few examples of RRIDs that should be autolinked)

File: apps/portals/e2e/configs/routesConfig.ts [hunk 3]
```diff
@@ -17,13 +17,7 @@ type RouteConfig = Record<Portal, RouteButtonsLinks>
...
   adknowledgeportal: {
     buttons: ['Home', 'Explore', 'Sign In'],
-    links: [
-      'Analytical Workspace',
-      'Data Access',
-      'Contribute',
-      'News',
-      'Help',
-    ],
+    links: ['Analysis Platforms', 'Data Access', 'Contribute', 'News', 'Help'],
```
Comment (jay-hodgson) [inline apps/portals/e2e/configs/routesConfig.ts:20]: I broke these tests in a previous PR

Comment (jay-hodgson) [reply apps/portals/e2e/configs/routesConfig.ts]: Hallie fixed in https://github.com/Sage-Bionetworks/synapse-web-monorepo/pull/634

File: packages/synapse-react-client/src/components/GenericCard/Linkify.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  regex: /((ftp|http|https):\/\/[^ ",]+)/,
+  onMatch: value => value,
+}
+
+const pubMedRule: LinkifyRule = {
+  regex: /(PMID:\d+)/,
+  onMatch: value => {
+    const id = value.slice('PMID:'.length)
+    return `https://pubmed.ncbi.nlm.nih.gov/${id}`
+  },
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/GenericCard/Linkify.tsx:31]: Could change the capture group so you don't have to `slice`, but it's up to you
```suggestion
  regex: /PMID:(\d+)/,
  onMatch: value => `https://pubmed.ncbi.nlm.nih.gov/${value}`
```

Comment (nickgros) [inline packages/synapse-react-client/src/components/GenericCard/Linkify.tsx:31]: Could change the capture group so you don't have to `slice`, but it's up to you
```suggestion
  regex: /PMID:(\d+)/,
  onMatch: value => `https://pubmed.ncbi.nlm.nih.gov/${value}`
```

File: packages/synapse-react-client/src/components/GenericCard/Linkify.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  regex:
+    /(AB_\d+)|(Addgene_\d+)|(CVCL_[a-zA-Z0-9]+)|(MMRRC_\d+-.{3})|(IMSR_JAX:\d+)/,
+  onMatch: value => `https://scicrunch.org/resolver/${value}`,
+}
+const clinVarVCVRule: LinkifyRule = {
+  regex: /(ClinVar:VCV\d+)/,
+  onMatch: value => {
+    const id = value.slice('ClinVar:'.length)
+    return `https://www.ncbi.nlm.nih.gov/clinvar/variation/${id}`
+  },
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/GenericCard/Linkify.tsx:47]: Ditto above for changing the capture group

File: packages/synapse-react-client/src/components/GenericCard/Linkify.tsx (hunk 4 above)
Comment (nickgros) [inline packages/synapse-react-client/src/components/GenericCard/Linkify.tsx:31]: Could change the capture group so you don't have to `slice`, but it's up to you
```suggestion
  regex: /PMID:(\d+)/,
  onMatch: value => `https://pubmed.ncbi.nlm.nih.gov/${value}`
```

File: packages/synapse-react-client/src/components/GenericCard/Linkify.tsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+}
+const clinVarRCVRule: LinkifyRule = {
+  regex: /(ClinVar:RCV\d+)/,
+  onMatch: value => {
+    const id = value.slice('ClinVar:'.length)
+    return `https://www.ncbi.nlm.nih.gov/clinvar/${id}`
+  },
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/GenericCard/Linkify.tsx:54]: And again

File: packages/synapse-react-client/src/components/GenericCard/Linkify.tsx [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  clinVarRCVRule,
+]
+const splitter = new RegExp(rules.map(r => r.regex.source).join('|'), 'g')
+
+const Linkify: React.FC<LinkifyProps> = ({ text, className }) => {
+  if (text === undefined) {
+    return <></>
+  } else if (!text.split) {
+    return <p>{text}</p>
+  }
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/GenericCard/Linkify.tsx:73]: Could this be simplified?
```suggestion
  if (text == null) {
    return <></>
  }
```

File: packages/synapse-react-client/src/components/GenericCard/Linkify.tsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+  const parts = text.split(splitter)
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/GenericCard/Linkify.tsx:75]: Took me a moment and had to look back at the tests to figure out why we needed to split here, so I think a quick note would be helpful
```suggestion
  // Split the string using the joined regular expressions so we can add links to identifiers within free text
  const parts = text.split(splitter)
```

File: packages/synapse-react-client/src/components/GenericCard/SynapseCardLabel.tsx [hunk 9]
```diff
@@ -97,23 +96,7 @@ export const SynapseCardLabel: React.FC<SynapseCardLabelProps> = props => {
...
-          className={newClassName}
-        >
-          {str}
-        </Link>
-      )
-    } else {
-      // they don't need a link
-      return <p>{str}</p>
-    }
+    return <Linkify text={str} className={newClassName} />
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/GenericCard/SynapseCardLabel.tsx:99]: Really nice encapsulation

File: packages/synapse-react-client/src/components/SynapseTable/SynapseTableCell/SynapseTableCell.tsx [hunk 10]
```diff
@@ -192,7 +193,7 @@ function SynapseTableCell(props: SynapseTableCellProps) {
           {jsonData.map((val: string, index: number) => {
             return (
               <React.Fragment key={val}>
-                {val}
+                <Linkify text={val} />
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/SynapseTable/SynapseTableCell/SynapseTableCell.tsx:196]: Failing test related to handling INTEGER_LIST here, you could probably either not `Linkify` INTEGER_LIST or change the test to be more permissive

**Review by nickgros — APPROVED** [review]
Looks good! Had optional suggestions, seems to be one test failure that is hopefully an easy fix

File: packages/synapse-react-client/src/components/GenericCard/Linkify.tsx (hunk 4 above)
Comment (nickgros) [reply packages/synapse-react-client/src/components/GenericCard/Linkify.tsx:31]: Your way is arguably more readable, so I'm ok with either

File: packages/synapse-react-client/src/components/GenericCard/Linkify.tsx (hunk 8 above)
Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/GenericCard/Linkify.tsx:73]: Added comment

**Pull request merged**